Реализация ФЗ-12.

Novacancy531 opened the pull request.

Novacancy531 committed:
Реализация ФЗ-12.

Novacancy531 committed:
Исправления:
1. Регистр data.sql
2. Переключение хранения базы данных на работу в памяти.

VadimZharkov commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,44 +1,56 @@
+    private final FilmStorage filmStorage;
+    private final UserStorage userStorage;
+
+    public FilmService(@Qualifier("filmRepository") final FilmStorage filmStorage,
```
Т.к. у нас только одна реализация FilmStorage (InMemory классы мы удалили), то сработает и без `@Qualifier`

VadimZharkov commented on src/main/java/ru/yandex/practicum/filmorate/storage/dal/FilmRepository.java:
```diff
@@ -0,0 +1,134 @@
+public class FilmRepository implements FilmStorage {
+
+    private final JdbcTemplate jdbcTemplate;
+    private final GenreRepository genreRepository;
```
У класса репозитория не должно быть ссылок на другие репозитории. Если нужно сделать проверки, то это зона ответственности класса-сервиса. Он композитный класс, т.е. в нем допустимы ссылки на другие репозитории и сервисы.

VadimZharkov commented on src/main/java/ru/yandex/practicum/filmorate/storage/dal/FilmRepository.java:
```diff
@@ -0,0 +1,134 @@
+        }
+        film.setId(key.longValue());
+
+        genreRepository.saveFilmGenres(film);
```
Вынести в класс-сервис

VadimZharkov commented on src/main/java/ru/yandex/practicum/filmorate/storage/dal/FilmRepository.java:
```diff
@@ -0,0 +1,134 @@
+
+    @Override
+    public Film addFilm(final Film film) {
+        mpaRepository.validateMpa(film);
```
Вынести в класс-сервис

VadimZharkov commented on src/main/java/ru/yandex/practicum/filmorate/storage/dal/FilmRepository.java:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    @Override
+    public Map<Long, Film> getAllFilms() {
```
Не понятно зачем Map, вполне можно и List отдать

VadimZharkov commented on src/main/java/ru/yandex/practicum/filmorate/storage/dal/FilmRepository.java:
```diff
@@ -0,0 +1,134 @@
+                film.getId()
+        );
+
+        genreRepository.deleteFilmsGenres(film.getId());
```
Вынести в класс-сервис

VadimZharkov commented on src/main/java/ru/yandex/practicum/filmorate/storage/dal/FilmRepository.java:
```diff
@@ -0,0 +1,134 @@
+        );
+
+        genreRepository.deleteFilmsGenres(film.getId());
+        genreRepository.saveFilmGenres(film);
```
Вынести в класс-сервис

VadimZharkov commented on src/main/java/ru/yandex/practicum/filmorate/storage/dal/FilmRepository.java:
```diff
@@ -0,0 +1,134 @@
+
+    @Override
+    public void deleteFilm(final long id) {
+        genreRepository.deleteFilmsGenres(id);
```
Вынести в класс-сервис

VadimZharkov commented on src/main/java/ru/yandex/practicum/filmorate/storage/dal/FilmRepository.java:
```diff
@@ -0,0 +1,134 @@
+    private final MPARepository mpaRepository;
+    private static final String ADD_FILM = "INSERT INTO films (name, description, release_date, duration, mpa_id)" +
+            " VALUES (?, ?, ?, ?, ?)";
+    private static final String FIND_FILM = "SELECT * FROM films WHERE film_id = ?";
```
Лучше заджойнить с mpa, чтобы в mapRowToFilm не делать доп запроса к mpa

VadimZharkov commented on src/main/java/ru/yandex/practicum/filmorate/storage/dal/FilmRepository.java:
```diff
@@ -0,0 +1,134 @@
+
+        film.setMpa(mpaRepository.getNameById(rs.getInt("mpa_id")));
+
+        Set<Genre> genres = genreRepository.findGenresByFilmId(film.getId());
```
В методах мапперах не нужно делать доп запросов, это лучше сделать в классе-сервисе.

VadimZharkov commented on src/main/java/ru/yandex/practicum/filmorate/storage/dal/GenreRepository.java:
```diff
@@ -0,0 +1,94 @@
+
+        if (!film.getGenres().isEmpty()) {
+            for (Genre genre : film.getGenres()) {
+                jdbcTemplate.update(SAVE_GENRE, film.getId(), genre.getId());
```
Вставлять по одной записи в цикле не оптимально.  Лучше воспользоваться пакетной операцией
https://habr.com/ru/articles/703828/

VadimZharkov reviewed (changes requested):
Нужно исправить замечания.

Novacancy531 committed:
Исправления:
1. Разделены репозитории для работы с фильмами, жанрами и рейтингами.
2. Возвратное значение метода получения всех фильмов заменено на SET.
3. Удалены InMemory классы.
4. Оптимизированы запросы к базе данных для сохранения жанров фильма.

VadimZharkov reviewed (approved):
Доработки увидел, принимается.
